improver/ensemble_copula_coupling/numba_utilities.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown Copyright 2017-2021 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+This module defines the optional numba utilities for Ensemble Copula Coupling
+plugins.
+
+"""
+
+import os
+
+import numpy as np
+from numba import config, njit, prange, set_num_threads
+
+config.THREADING_LAYER = "omp"
+if "OMP_NUM_THREADS" in os.environ:
+    set_num_threads(int(os.environ["OMP_NUM_THREADS"]))
+
+
+@njit(parallel=True)
+def fast_interp_same_y(x: np.ndarray, xp: np.ndarray, fp: np.ndarray) -> np.ndarray:
+    """For each row i of xp, do the equivalent of np.interp(x, xp[i], fp).
+    Args:
+        x: 1-d array
+        xp: n * m array, each row must be in non-decreasing order
+        fp: 1-d array with length m
+    Returns:
+        n * len(x) array where each row i is equal to np.interp(x, xp[i], fp)
+    """
+    # check whether x is non-decreasing
+    x_ordered = True
+    for i in range(1, len(x)):
+        if x[i] < x[i - 1]:
+            x_ordered = False
+            break
```

Comment:
As an alternative to checking the order and handling it differently throughout, I think you can just sort the input and reorder the output:
```suggestion
    xorder = np.argsort(x)
    xorig = np.empty_like(xorder)
    xorig[xorder] = np.arange(xorder.size)
    x = x[xorder]
```

This would then require your return statement to be:
```
return result[:, xorig]
```

It would allow you to remove the handling for ordered vs disordered. This PR and that for the percentile resampling essentially share the same actual steps to perform the interpolation (though written a bit differently), so that could be factored out and shared between two wrappers for the probability or percentile inputs.

Reply:
Thanks, this is a good idea!

Reply:
Actually, the re-ordering seems to make it significantly slower (by a factor of > 2), so I have restored the old code here.

Reply:
Regarding factoring out common code, I actually don't think there's much opportunity for this. The code for a single row of the result is different in each case because in this PR `xp` is 2-dimensional, and in the other PR, it's `fp`, so all the indexing is a bit different. Also, in this PR, we have 2 different code paths depending on whether `x` is sorted. This avoids calling `np.searchsorted` once per row of `xp`, so gives us an efficiency gain. In the other PR, however, there's only one call of `np.searchsorted`, so there's no opportunity for this speedup.

Reply:
If it makes is slower then that does rather undermine the purpose, so fair enough. As to the factoring out, yep, it would probably add more complexity trying to make them share code so it's also fine to leave these as seperate. Sorry for sending you down a rabbit hole.